### DISCUSSION TITLE
How to use Client Extensions to manage types w/ TypeScript & Postgresql

### DISCUSSION THREAD

**Discussion by linjer** [prisma/prisma#16974] [Q&A]
Hello!

I was initially excited to learn about the addition of the Client Extensions preview feature because our code (using TypeScript and Postgresql) needs to post process records often.

For example, lets say there is an `Account` model, where each record may have a `metadata` field that has key-value pairs. This is stored as a `JsonObject` but when we pull it from the DB we'd like to be casted into the `AccountMetadata` object type.

Another example is when handling money, we may store the amount as an integer (cents) in the DB, but we would like to wrap it in a money type when we pull it from the DB.

The documentation for the Client Extensions seems describe a way to add new computed fields, but no way to mutate and replace existing fields. Replacement would be preferred to avoid confusing bugs using the similar but not-typed/wrapped object. Our current approach is to write and maintain a special abstraction layer that converts into a companion type, but it gets messy, gets tricky if we want to put multiple operations into a `$transaction`, and doesn't prevent people from just accessing the prisma client directly.

Is there a way to do this with Client Extensions or otherwise?

**Comment by pantharshit00** [comment]
Hello,

Client extensions will also allow you to mutate existing fields but right now we have a bug in the implementation which we are working on fixing: https://github.com/prisma/prisma/issues/16643

You can also refer to this example which I think covers your case here. But right now if it casting data to a new field which you can change once the above mentioned bug is fixed: https://github.com/sbking/prisma-client-extensions/blob/main/examples/json-field-types/src/index.ts

**Comment by knnedy** [comment]
You can use Zod its great for Type safety.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
